trace-utils/src/no_alloc_string.rs

```diff
@@ -0,0 +1,149 @@
+use serde::ser::{Serialize, Serializer};
+use std::borrow::Borrow;
+use tinybytes;
+
+pub struct BufferWrapper {
+    buffer: tinybytes::Bytes,
+}
+
+impl BufferWrapper {
+    pub fn new(buffer: tinybytes::Bytes) -> Self {
+        BufferWrapper { buffer }
+    }
+
+    /// Creates a NoAllocString from a slice of tinybytes::Bytes.
+    pub fn create_no_alloc_string(
+        &self,
+        slice: &[u8],
+    ) -> Result<NoAllocString, std::str::Utf8Error> {
+        NoAllocString::from_bytes(self.buffer.slice_ref(slice).expect("Invalid slice"))
+    }
+
+    /// Creates a NoAllocString from a slice of tinybytes::Bytes without validating the bytes are
+    /// UTF-8.
+    pub fn create_no_alloc_string_unchecked(&self, slice: &[u8]) -> NoAllocString {
+        NoAllocString::from_bytes_unchecked(self.buffer.slice_ref(slice).expect("Invalid slice"))
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct NoAllocString {
+    bytes: tinybytes::Bytes,
+}
+
+impl Serialize for NoAllocString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.as_str())
+    }
+}
+
+impl NoAllocString {
+    // Creates a NoAllocString from a full slice (copies the data, so technically not no-alloc)
+    pub fn from_slice(slice: &[u8]) -> Result<NoAllocString, std::str::Utf8Error> {
+        std::str::from_utf8(slice)?;
+        Ok(NoAllocString {
+            bytes: tinybytes::Bytes::copy_from_slice(slice),
+        })
+    }
+    // Creates a NoAllocString from a Bytes instance (does not copy the data)
+    pub fn from_bytes(bytes: tinybytes::Bytes) -> Result<NoAllocString, std::str::Utf8Error> {
+        std::str::from_utf8(&bytes)?;
+        Ok(NoAllocString { bytes })
+    }
+
+    /// Creates a NoAllocString from a Bytes instance without validating the bytes are UTF-8.
+    pub fn from_bytes_unchecked(bytes: tinybytes::Bytes) -> NoAllocString {
```

Comment:
The proper way would be probably to mark this unsafe and assert or assume safety on caller sites.